Improve IncrementalIndex concurrency scalability

eranmeir opened the pull request:
Minimizing the critical section in adding input rows:
  - Using ConcurrentHashMap for dimensionDescs, updating atomically using computeIfAbsent
  - Duplicate dimension detection in a single row is now done using a local hashmap, which reduces global state access and also makes overflow array unnecessary.
  - getDimensions() now returns an immutable copy of dimensionDescsList to keep order requirements (since ConcurrentHashMap values are not ordered)
- Adding multithreaded benchmark

eranmeir committed:
Improve IncrementalIndex concurrency scalability

- Shared state is encapsulated in a new class - DimensionData. This includes dimensionDescs, dimensionDescsList and columnCapabilities
- Concurrent threads share an atomic reference to an instance of DimensionData
- CoW: Only when a thread needs to update the shared state, it will copy the instance, update the copy, and eventually swap the reference atomically.
- Consistency is maintained when the reference is updated. This simplifies row processing, removes the need for keeping an “overflow” array, and allows fast failure when a row contains duplicate dimensions.
- New multi-threaded ingestion benchmark: IndexIngestionMultithreadedBenchmark

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
+        capabilities = columnCapabilities.get(dimension);
+        if (capabilities == null) {
+          capabilities = new ColumnCapabilitiesImpl();
+          // For schemaless type discovery, assume everything is a String for now, can change later.
```
Not clear, is this a special case? Do we need more changes prior to publishing?

Eshcar replied:
most of this code did not change - it is just different indentation

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
+          capabilities.setType(ValueType.STRING);
+          capabilities.setDictionaryEncoded(true);
+          capabilities.setHasBitmapIndexes(true);
+          columnCapabilities.putIfAbsent(dimension, capabilities);  // eranmeir - can skip the above null checking as it is done by putIfAbsent
```
What happens if putIfAbsent() fails? Can you still use the capabilities object?

Eshcar replied:
I wouldn't skip the null check; we want to avoid contention on the shared data structure if possible

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
+          columnCapabilities.putIfAbsent(dimension, capabilities);  // eranmeir - can skip the above null checking as it is done by putIfAbsent
         }
+        DimensionHandler handler = DimensionHandlerUtils.getHandlerFromCapabilities(dimension, capabilities, null);
+        desc = addNewDimension(dimension, capabilities, handler);
```
I am afraid there is a race condition here. If putIfAbsent() fails, the code should use the new value of capabilities.

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -824,24 +797,18 @@ private long getMaxTimeMillis()
-    synchronized (dimensionDescs) {
-      return ImmutableList.copyOf(dimensionDescs.keySet());
-    }
+    return ImmutableList.copyOf(dimensionDescs.keySet());
```
Apparently the original code had to make a copy because dimensionDescs is not thread safe. Do we have to make it here?

eranmeir replied:
I tried looking for usages of this method and there don't seem to be any that mutate the list, so I guess it would be safe. That been said, we would probably have to linearly scan/copy the keyset  to transform it to a list, so defensively making it immutable is probably a good practice.

Also, Google [claims](https://github.com/google/guava/wiki/ImmutableCollectionsExplained#copyof-is-smarter-than-you-think) that `copyOf` is smarter than naive copy, although I'm not sure if that's true for copies of mutable objects.

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
-            handler.getLengthOfEncodedKeyComponent(dimsKey) > 1) {
-          capabilities.setHasMultipleValues(true);
-        }
+    for (String dimension : rowDimensions) {
```
This method is very long. I know that's legacy, but some refactoring could have greatly improved readability.

ebortnik commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -925,28 +890,27 @@ public void loadDimensionIterable(
-    dimensionDescs.put(dim, desc);
-    dimensionDescsList.add(desc);
-    return desc;
+    return dimensionDescs.computeIfAbsent(dim, d -> {
```
Let's discuss this ... Not sure I understood.

sanastas commented on benchmarks/src/main/java/org/apache/druid/benchmark/indexing/IndexIngestionMultithreadedBenchmark.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  @Benchmark
+  @BenchmarkMode(Mode.SingleShotTime)
```
I wonder whether this is the best metric for multi-threading... How hard/correct would be to switch to throughput? A total counter of inserted rows?

eranmeir replied:
The original motivation to use singleshot was because other modes just repeat the benchmark as many times as possible in a given time frame. For ingestion, this basically means re-inserting existing keys, which skews the benchmark.
Perhaps we can think of a different way to implement the benchmark that can work around this.

sanastas replied:
If a key is re-inserted (both on Oak and native Druid index) the compute is done. Which can also be something to measure... meaning throughput...

sanastas commented on benchmarks/src/main/java/org/apache/druid/benchmark/indexing/IndexIngestionMultithreadedBenchmark.java:
```diff
@@ -0,0 +1,158 @@
+    Options opt = new OptionsBuilder()
+        .include(IndexIngestionMultithreadedBenchmark.class.getSimpleName())
+        .forks(1)
+        .threads(4)
```
Can you change number of threads from the command line? In general, the command line parameters (of how to invoke the benchmark) are important. Can we add some comment of how to invoke the test and what to expect?

eranmeir replied:
Yes, you can run the benchmark with  `-t [number]` to run with multiple threads.
The command for running the multithreaded benchmark is basically:
```
java -jar pre-fix.jar IndexIngestionMultithreadedBenchmark -p rowsPerSegment=75000  -p schema=basic -t 2
```
(replacing schema/thread count as necessary)

sanastas replied:
Maybe to mention it somewhere in the code, so it can be clearer how to use the benchmark

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
+    Map<Integer, Object> rowDims = new HashMap<>();
     long dimsKeySize = 0;
     List<String> parseExceptionMessages = new ArrayList<>();
-    synchronized (dimensionDescs) {
```
since you removed this line the indentation changed and it is harder to see the core modifications

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
     final List<String> rowDimensions = row.getDimensions();
 
-    Object[] dims;
-    List<Object> overflow = null;
```
we need to explain how can we give up the overflow array

eranmeir replied:
I think it's obvious that we now use a local map in the code. It's also explained in the commit message (and the PR description), FWIW.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -630,91 +629,65 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
-      for (int i = 0; i < overflow.size(); ++i) {
-        newDims[dims.length + i] = overflow.get(i);
+      if (rowDims.putIfAbsent(desc.getIndex(), dimsKey) != null) {
+        // If the dims map already contains a mapping at this index, it means we have seen this dimension already on this input row.
```
why?
i don't get it - 
first can't it be that it was inserted by a different row?
second, didn't we say we'll check the validity of the row before processing it, and throw the exception at the beginning. we don't want to process an invalid row

eranmeir replied:
> can't it be that it was inserted by a different row?

No. `rowDims` is a local map, which is used exactly to track unique dimensions in the current row.

> didn't we say we'll check the validity of the row before processing it, and throw the exception at the beginning.

Yes, we can do this. Will fix.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -925,28 +890,27 @@ public void loadDimensionIterable(
+        ColumnCapabilitiesImpl capabilities = oldColumnCapabilities.get(dim);
+        columnCapabilities.put(dim, capabilities);
+        DimensionHandler handler = DimensionHandlerUtils.getHandlerFromCapabilities(dim, capabilities, null);
+        addNewDimension(dim, capabilities, handler);
```
I don't see why removing the synchronizes will work. it is currently protecting 2 hash maps that can be modified concurrently

eranmeir replied:
`columnCapabilities` appears to hold mapping that are, more or less, constant. It seems to me like it doesn't matter who populates a mapping first, this just saves extra instantiations, so putIfAbsent seems to be enough for this purpose.

I will look again to make sure.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -925,28 +890,27 @@ public void loadDimensionIterable(
-    return desc;
+    return dimensionDescs.computeIfAbsent(dim, d -> {
+      // Since these two structures are populated together atomically, we can use dimensionDescsList.size(), which is faster.
+      DimensionDesc desc = new DimensionDesc(dimensionDescsList.size(), dim, capabilities, handler);
```
but they are not populated together atomically, at least I can't see why you think they are

eranmeir replied:
They are populated together in the sense that the list matches the map's values. I meant that atomicity here comes from `ConcurrentHashMap`'s `computIfAbsent`

Eshcar reviewed (commented):
The synchronize lines that were removed protect modification of two hash maps 
it is not sufficient to swap them with concurrent hash maps to maintain their atomic modification.
we can think of some copy-on-write mechanism that is protected either by a version number to make sure  readers see a consistent view and apply changes atomically

this can also help with improving the performance of contention-free runs

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -925,30 +919,20 @@ public void loadDimensionIterable(
-          DimensionHandler handler = DimensionHandlerUtils.getHandlerFromCapabilities(dim, capabilities, null);
-          addNewDimension(dim, capabilities, handler);
-        }
+    DimensionData dimensionData = dimensions.get().clone();
```
how often is this method invoked? if it is often then maybe clone only if needed (i.e. size != 0)

eranmeir replied:
I don't think it is called often. From what I understand, this is only called during index creation, and only in certain cases.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -824,24 +823,20 @@ private long getMaxTimeMillis()
-    synchronized (dimensionDescs) {
-      return ImmutableList.copyOf(dimensionDescs.values());
-    }
+    return dimensions.get().clone().getDimensionDescsList();
```
are we sure cloning is always needed here?
If the method is invoked only for reading the list then cloning is redundant and expensive.
better to clone explicitly only when needed, 
can wrap it with a different method

eranmeir replied:
The original method returned an immutable copy of dimension descriptors. I looked at callers of this method and it seems like they don't mutate it, so I guess it's ok.

sanastas replied:
According to what Eran says if the callers doesn't mutate and only read, it is OK not to clone... Can we try that?

sanastas commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -311,21 +306,23 @@ protected IncrementalIndex(
       this.rowTransformers.add(new SpatialDimensionRowTransformer(spatialDimensions));
     }
+
+    this.dimensions = new AtomicReference<>(dimensionData);
```
What are the assumptions on the thread-safeness of the constructor? I wonder whether it is 100% safe to make assignment without any synchronization? Do we assume

eranmeir replied:
Not sure I understand the question. If the object is not yet constructed, threads shouldn't be able to call any (non-static) methods.

eranmeir committed:
Fixing dimension order in index creation

Added missing CAS operation after dimension updates

eranmeir committed:
Fixing some code review issues (WIP)

- Permissive duplicate dimension handling (tests updated)
- Additional error logging

eranmeir committed:
Fixing some code review issues (WIP)

- Refactoring toIncrementalIndexRow
- Additional comments indicating code that should be thread-safe
